src/view3d.cpp

```diff
@@ -185,61 +209,156 @@ void View3D::centerOnGeometry(const QModelIndex& index)
+    // It also doesn't fetch the function pointers. The solution is for GLEW to use glGetStringi​ instead.
+    // The current version of GLEW is 1.10.0 but they still haven't corrected it.
+    // The only fix is to use glewExperimental​ for now.
+    // https://www.opengl.org/wiki/OpenGL_Loading_Library
+    glewExperimental = true;
     if (glewInit() != GLEW_OK)
     {
         g_logger.error("%s", "Failed to initialize GLEW");
         m_badOpenGL = true;
         return;
     }
+
     g_logger.info("OpenGL implementation:\n"
-                  "GL_VENDOR   = %s\n"
-                  "GL_RENDERER = %s\n"
-                  "GL_VERSION  = %s",
+                  "GL_VENDOR    = %s\n"
+                  "GL_RENDERER  = %s\n"
+                  "GL_VERSION   = %s\n"
+                  "GLSL_VERSION = %s\n"
+                  "GLEW_VERSION = %s\n",
                   (const char*)glGetString(GL_VENDOR),
                   (const char*)glGetString(GL_RENDERER),
-                  (const char*)glGetString(GL_VERSION));
-    m_shaderProgram->setContext(context());
-    m_meshFaceShader.reset(new ShaderProgram(context()));
+                  (const char*)glGetString(GL_VERSION),
+                  (const char*)glGetString(GL_SHADING_LANGUAGE_VERSION),
+                  (const char*)glewGetString(GLEW_VERSION));
+
+    // GL_CHECK has to be defined for this to actually do something
+    glCheckError();
+
+    initCursor(10, 1);
+    initAxes();
+    initGrid(2.0f);
+
+    glCheckError();
+
+    m_boundingBoxShader.reset(new ShaderProgram());
+    m_boundingBoxShader->setShaderFromSourceFile("shaders:bounding_box.glsl");
+
+    m_meshFaceShader.reset(new ShaderProgram());
     m_meshFaceShader->setShaderFromSourceFile("shaders:meshface.glsl");
-    m_meshEdgeShader.reset(new ShaderProgram(context()));
+
+    m_meshEdgeShader.reset(new ShaderProgram());
     m_meshEdgeShader->setShaderFromSourceFile("shaders:meshedge.glsl");
-    m_incrementalFramebuffer = allocIncrementalFramebuffer(width(), height());
+
+    double dPR = getDevicePixelRatio();
+    int w = width() * dPR;
+    int h = height() * dPR;
+
+    m_incrementalFramebuffer = allocIncrementalFramebuffer(w, h);
+
+    glFrameBufferStatus(m_incrementalFramebuffer);
+    glCheckError();
+
     const GeometryCollection::GeometryVec& geoms = m_geometries->get();
     for (size_t i = 0; i < geoms.size(); ++i)
-        geoms[i]->initializeGL();
+    {
+        if (m_boundingBoxShader->isValid())
+        {
+            // TODO: build a shader manager for this
+            geoms[i]->setShaderId("boundingbox", m_boundingBoxShader->shaderProgram().programId());
+            geoms[i]->setShaderId("meshface", m_meshFaceShader->shaderProgram().programId());
+            geoms[i]->setShaderId("meshedge", m_meshEdgeShader->shaderProgram().programId());
+            geoms[i]->initializeGL();
+        }
+    }
+
+    PointViewerMainWindow * pv_parent = dynamic_cast<PointViewerMainWindow *>(parentWidget());
```

Comment:
Oh dear, so this is why you needed initializedGL()... so that the shader editor would also get the correct shader without something horrible like this `dynamic_cast`.  Sorry about that... can be cleaned up later.  Looks like I should connect up the text editor's `QTextDocument` to the shader or some such thing - some kind of more direct connection rather than proxying through the `PointViewMainWindow`...